utils/MaxHeap.ts

```diff
@@ -0,0 +1,80 @@
+export default class MaxHeap<T> {
+  private heap: T[];
+  private compare: (node1: T, node2: T) => 1 | -1 | 0;
+  /**
+   *
+   * @param compareFunction return 1 if node 1 > node 2, return -1 if node 1 < node 2, return 0 if node1 = node2
+   */
+  constructor(compareFn: (node1: T, node2: T) => 1 | -1 | 0) {
+    this.heap = [];
+    this.compare = compareFn;
+  }
+
+  get length(): number {
+    return this.heap.length;
+  }
+
+  public peek() {
+    if (this.length > 0) return this.heap[0];
+    else return undefined;
+  }
+
+  public push(node: T) {
+    this.heap.push(node);
+    this.bubbleUp();
+  }
+
+  public pop(): T | undefined {
+    if (this.length === 0) {
+      return undefined;
+    }
+    [this.heap[0], this.heap[this.length - 1]] = [
+      this.heap[this.length - 1],
+      this.heap[0]
+    ];
+    const pop = this.heap.pop();
+    this.bubbleDown();
+    return pop;
+  }
+
+  private bubbleUp() {
+    let current = this.length - 1;
+    while (current > 0) {
+      let parent = Math.ceil(current / 2) - 1;
+      if (this.compare(this.heap[current], this.heap[parent]) == 1) {
+        [this.heap[current], this.heap[parent]] = [
+          this.heap[parent],
+          this.heap[current]
+        ];
+        current = parent;
+      } else {
+        return;
+      }
+    }
+  }
+
+  private bubbleDown() {
```

Comment:
if we gonna define the return type for these methods, I think the implementation will be much more elegant